geomesa-fs/geomesa-fs-storage-parquet/src/main/scala/org/locationtech/geomesa/parquet/ParquetFileSystemStorage.scala

```diff
@@ -222,15 +229,116 @@ class ParquetFileSystemStorage(root: Path,
+    var tryNum = 0
+    var backupComplete = false
+
+    def waitOnBackup: Boolean = {
+      backupComplete = fs.exists(backupFile) && !fs.exists(metaPath)
+      if (!backupComplete) {
+        val secs = 2 ^ tryNum
+        Thread.sleep(1000 * secs)
+      }
+      !backupComplete
+    }
+
+    do {
+      tryNum += 1
+    } while (waitOnBackup && tryNum <= 3)
+
+    if (!backupComplete) {
+      throw new IOException(s"Unable to properly backup metadata after $tryNum tries")
+    }
+  }
+
   override def updateMetadata(typeName: String): Unit = {
     val s = System.currentTimeMillis
     val scheme = metadata(typeName).getPartitionScheme
-    val parts = StorageUtils.partitionsAndFiles(root, fs, typeName, scheme, StorageUtils.SequenceLength, FileExtension)
+    val sft = metadata(typeName).getSimpleFeatureType
+    val parts = StorageUtils.partitionsAndFiles(root, fs, typeName, scheme, FileExtension)
+
+    // Save existing metadata
+    backupMetadata(typeName)
+    cleanBackups(typeName)
+
+    // Recreate a new metadata file
+    val newMetadata = createFileMetadata(sft)
+    MetadataCache.invalidate((root, typeName))
+    MetadataCache.put((root, typeName), newMetadata)
+
     metadata(typeName).addPartitions(parts)
     val e = System.currentTimeMillis
     logger.info(s"Metadata Update took in ${e-s}ms.")
   }
 
+  override def compact(typeName: String, partition: String): Unit = {
+    val existingFiles = getPaths(typeName, partition)
+
+    val meta = metadata(typeName)
+    val sft = meta.getSimpleFeatureType
+
+    val sftConf = {
+      val c = new Configuration(conf)
+      SimpleFeatureReadSupport.setSft(sft, c)
+      c
+    }
+    val leaf = meta.getPartitionScheme.isLeafStorage
+    val dataPath = StorageUtils.nextFile(fs, root, typeName, partition, leaf, FileExtension, FileType.Compacted)
+    val writer = SimpleFeatureParquetWriter.builder(dataPath, sftConf).build()
+
+    logger.debug(s"Compacting data files: [${existingFiles.map(_.toString).mkString(", ")}] to into file $dataPath")
+    val support = new SimpleFeatureReadSupport
+    val written: Long = existingFiles.map { f =>
+      logger.debug(s"Reading $f")
+      val reader = ParquetReader.builder[SimpleFeature](support, new Path(f)).withConf(sftConf).build()
+      var sf = reader.read()
+      var count = 0L
+      while (sf != null) {
+        writer.write(sf)
+        count += 1
+        sf = reader.read()
+      }
+      count
+    }.sum
+
+    writer.close()
+    logger.debug(s"Wrote compacted file $dataPath")
+
+    logger.debug(s"Deleting old files [${existingFiles.map(_.toString).mkString(", ")}]")
+    val deleteResult = existingFiles.forall(f => fs.delete(new Path(f), false))
+    if (!deleteResult) {
+      logger.warn(s"Failed to delete all files: [${existingFiles.map(_.toString).mkString(", ")}]")
+    }
+
+    logger.debug(s"Updating metadata for type $typeName")
+    updateMetadata(typeName)
```

Comment:
not sure how all this is synchronized - does it matter that files are deleted before metadata is updated? seems like metadata would be invalid while files are deleting and before it was updated